### PR TITLE
Reciprocal-based 2by1 division procedure

### PR DESCRIPTION
**Pull request by chfast**
This is based on [Improved division by invariant integers](https://gmplib.org/~tege/division-paper.pdf) and ported from https://github.com/chfast/intx.

The computation of reciprocal is single `DIV` x86_64 instruction, but there exist faster implementation (sneak peek in #39) but I left it for other time.

```
name                     old time/op  new time/op  delta
_Div/large/big-6          246ns ± 0%   245ns ± 0%   -0.49%  (p=0.016 n=4+5)
_Div/large/uint256-6     99.2ns ± 0%  83.3ns ± 0%  -16.07%  (p=0.008 n=5+5)
_Div/small/big-6         68.8ns ± 0%  68.8ns ± 0%     ~     (p=0.500 n=5+5)
_Div/small/uint256-6     13.1ns ± 0%  13.0ns ± 0%   -0.76%  (p=0.008 n=5+5)
_MulMod/large/big-6       388ns ± 0%   388ns ± 0%     ~     (p=0.167 n=5+5)
_MulMod/large/uint256-6   204ns ± 0%   153ns ± 0%  -25.00%  (p=0.008 n=5+5)
_MulMod/small/big-6      94.1ns ± 1%  93.9ns ± 0%     ~     (p=0.397 n=5+5)
_MulMod/small/uint256-6  40.0ns ± 0%  40.0ns ± 0%     ~     (p=0.444 n=5+5)
MulMod/mod64-6            186ns ± 0%   101ns ± 0%  -45.70%  (p=0.008 n=5+5)
MulMod/mod128-6           266ns ± 0%   176ns ± 0%  -33.68%  (p=0.000 n=4+5)
MulMod/mod192-6           245ns ± 0%   173ns ± 0%     ~     (p=0.079 n=4+5)
MulMod/mod256-6           223ns ± 0%   169ns ± 0%  -24.22%  (p=0.000 n=5+4)
_Mod/large/big-6          137ns ± 0%   137ns ± 0%     ~     (all equal)
_Mod/large/uint256-6     50.2ns ± 0%  68.2ns ± 0%  +35.94%  (p=0.016 n=4+5)
_Mod/small/big-6         40.8ns ± 0%  40.9ns ± 0%     ~     (p=0.325 n=5+5)
_Mod/small/uint256-6     14.0ns ± 0%  14.0ns ± 0%     ~     (all equal)
_SDiv/large/big-6         402ns ± 0%   401ns ± 0%   -0.25%  (p=0.016 n=4+5)
_SDiv/large/uint256-6     105ns ± 0%    89ns ± 0%  -14.95%  (p=0.000 n=5+4)
_AddMod/large/big-6       212ns ± 0%   212ns ± 0%     ~     (all equal)
_AddMod/large/uint256-6  39.5ns ± 0%  39.9ns ± 1%   +0.96%  (p=0.024 n=5+5)
_AddMod/small/big-6      65.4ns ± 0%  65.3ns ± 1%     ~     (p=0.238 n=5+5)
_AddMod/small/uint256-6  17.2ns ± 0%  17.4ns ± 0%   +1.16%  (p=0.008 n=5+5)
```

### CODE REVIEW

**Comment by holiman** [comment]
Cool. But this stood out: 
```
_Mod/large/uint256-6     50.2ns ± 0%  68.2ns ± 0%  +35.94%  (p=0.016 n=4+5)
```
What happened there - any idea?

**Comment by chfast** [comment]
> Cool. But this stood out:
> 
> ```
> _Mod/large/uint256-6     50.2ns ± 0%  68.2ns ± 0%  +35.94%  (p=0.016 n=4+5)
> ```
> 
> What happened there - any idea?

I noticed it too late. Nevertheless, I can explain it. The numbers are suspicious anyway if you compare with `_Div/large/uint256-6     99.2ns ± 0%  83.3ns ± 0%  -16.07%` (exactly the same algo).

The benchmark is `0xfefefefefefefefe... mod 0xefefefefefefefef...`. The divisor is already normalized. The Knuth division requires only single step where it divides `<0, 0xfefefefefefefefe> / 0xefefefefefefefef` using single `DIV` instruction.

After the change we always compute the reciprocal, i.e. `<^d, 0xffffffffffffffff> / d` using the same single `DIV` instruction. Then any later division is computed by multiplication + additional adjustments.

So this is more work in this case for the new implementation. But there is more.

If you check the `DIV` instruction it is actually some kind of algorithm: 36 microinstructions, 32-96 cycles (Haswell), 35-88 cycles (Skylake). And when the numerator high part is zero (as when executing the benchmark in old implementation) it finishes quicker (verified empirically some time ago).

In summary, division is tricky because of many special cases in implementation. The same benchmarking it is tricky. That's why I prefer to generate multiple random samples and benchmark full samples set. Also: the complexity is `O(len(numerator) - len(divisor) + 1)` so 256 by 256 is not the worst case. And division by single word is implemented differently, division by two words _can_ be implemented differently.

I can rework benchmarking to do the same as in new `BenchmarkMulMod` using 1024 samples. This PR can be merged after.

**Comment by chfast** [comment]
More detailed Mod benchmarks in #42.

**Comment by chfast** [comment]
Updated benchmarks. The 256by256 division is slower by ~14%, the reason explained above^.
```
name                     old time/op  new time/op  delta
Div/small/uint256-6      13.3ns ± 0%  13.1ns ± 0%   -1.50%  (p=0.008 n=5+5)
Div/small/big-6          43.4ns ± 0%  43.4ns ± 0%     ~     (p=1.000 n=5+4)
Div/mod64/uint256-6      89.4ns ± 0%  57.3ns ± 0%  -35.91%  (p=0.000 n=5+4)
Div/mod64/big-6           105ns ± 0%   105ns ± 0%     ~     (all equal)
Div/mod128/uint256-6      112ns ± 0%    89ns ± 0%  -20.29%  (p=0.008 n=5+5)
Div/mod128/big-6          228ns ± 0%   228ns ± 1%     ~     (p=0.167 n=5+5)
Div/mod192/uint256-6     85.5ns ± 0%  79.8ns ± 0%   -6.76%  (p=0.008 n=5+5)
Div/mod192/big-6          193ns ± 0%   192ns ± 0%     ~     (p=0.079 n=5+5)
Div/mod256/uint256-6     59.2ns ± 0%  67.5ns ± 0%  +14.02%  (p=0.008 n=5+5)
Div/mod256/big-6          160ns ± 0%   159ns ± 0%   -0.62%  (p=0.029 n=4+4)
Mod/small/uint256-6      15.0ns ± 0%  14.9ns ± 2%     ~     (p=0.143 n=4+5)
Mod/small/big-6          42.3ns ± 0%  42.1ns ± 0%   -0.47%  (p=0.048 n=5+5)
Mod/mod64/uint256-6      90.6ns ± 0%  59.1ns ± 0%  -34.77%  (p=0.000 n=5+4)
Mod/mod64/big-6           115ns ± 0%   116ns ± 0%   +0.87%  (p=0.008 n=5+5)
Mod/mod128/uint256-6      113ns ± 0%    91ns ± 0%  -19.42%  (p=0.008 n=5+5)
Mod/mod128/big-6          226ns ± 0%   226ns ± 0%     ~     (all equal)
Mod/mod192/uint256-6     87.3ns ± 0%  81.7ns ± 0%   -6.39%  (p=0.008 n=5+5)
Mod/mod192/big-6          188ns ± 0%   189ns ± 0%   +0.53%  (p=0.029 n=4+4)
Mod/mod256/uint256-6     61.2ns ± 0%  69.7ns ± 0%  +13.95%  (p=0.008 n=5+5)
Mod/mod256/big-6          152ns ± 0%   149ns ± 1%   -1.71%  (p=0.016 n=4+5)
AddMod/small/uint256-6   18.4ns ± 0%  18.4ns ± 0%     ~     (all equal)
AddMod/small/big-6       59.2ns ± 0%  59.2ns ± 0%     ~     (p=1.000 n=5+5)
AddMod/mod64/uint256-6    196ns ± 0%   137ns ± 0%  -30.10%  (p=0.008 n=5+5)
AddMod/mod64/big-6        148ns ± 0%   147ns ± 0%   -0.68%  (p=0.000 n=5+4)
AddMod/mod128/uint256-6   245ns ± 0%   211ns ± 0%  -13.88%  (p=0.000 n=5+4)
AddMod/mod128/big-6       269ns ± 1%   267ns ± 0%   -0.74%  (p=0.024 n=5+5)
AddMod/mod192/uint256-6   200ns ± 0%   196ns ± 0%   -1.80%  (p=0.008 n=5+5)
AddMod/mod192/big-6       234ns ± 0%   233ns ± 0%     ~     (p=0.095 n=4+5)
AddMod/mod256/uint256-6   138ns ± 0%   157ns ± 0%  +13.77%  (p=0.008 n=5+5)
AddMod/mod256/big-6       198ns ± 1%   197ns ± 0%     ~     (p=0.079 n=5+5)
MulMod/small/uint256-6   41.2ns ± 0%  41.0ns ± 0%   -0.49%  (p=0.000 n=5+4)
MulMod/small/big-6       61.8ns ± 0%  61.9ns ± 0%     ~     (p=0.167 n=5+5)
MulMod/mod64/uint256-6    189ns ± 0%   101ns ± 0%  -46.56%  (p=0.008 n=5+5)
MulMod/mod64/big-6        242ns ± 0%   242ns ± 0%     ~     (all equal)
MulMod/mod128/uint256-6   266ns ± 0%   174ns ± 0%  -34.59%  (p=0.008 n=5+5)
MulMod/mod128/big-6       438ns ± 0%   437ns ± 0%     ~     (p=1.000 n=5+5)
MulMod/mod192/uint256-6   246ns ± 0%   172ns ± 0%  -30.08%  (p=0.008 n=5+5)
MulMod/mod192/big-6       402ns ± 0%   402ns ± 0%     ~     (p=0.556 n=4+5)
MulMod/mod256/uint256-6   229ns ± 0%   167ns ± 0%     ~     (p=0.079 n=4+5)
MulMod/mod256/big-6       374ns ± 0%   374ns ± 0%     ~     (p=1.000 n=5+5)
_SDiv/large/big-6         413ns ± 0%   416ns ± 0%   +0.73%  (p=0.008 n=5+5)
_SDiv/large/uint256-6     107ns ± 0%    89ns ± 0%  -16.54%  (p=0.000 n=5+4)
```
